File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if nums is None or len(nums) == 0:
            return -1
        
        low = 0
        high = len(nums) - 1

        while low <= high:
            mid = low + (high - low) // 2

            if nums[mid] == target:
                return mid

            if nums[low] <= nums[mid]:
                if nums[low] <= target <= nums[mid]:
                    high = mid - 1
                else:
                    low = mid + 1
            else:
                if nums[mid] <= target <= nums[high]:
                    low = mid + 1
                else:
                    high = mid - 1

        return -1

# Binary Search
#Time Complexity: O(logn)
#Space Complexity: O(1) 

#  [5,   1,    3]
# Low    mid   high

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        if(nums.length == 0 || nums == null){
            return -1;
        }
        
        int index = 0;
        
        for(int i = 1; i < nums.length; i++){
            if(nums[i] < nums[i - 1]){
                index = i;
                break;
            }
        }
        int low = -1;
        int high = -1;
        if (nums[index] <= target && target <= nums[nums.length - 1]){
             low = index;
             high = nums.length - 1;
        }
        else{
             low = 0;
             high = index - 1;
        }
        
        while (low <= high){
            int mid = low + (high - low) / 2;
            
            if(nums[mid] == target){
                return mid;
            }
            else if(nums[mid] < target){
                low = mid + 1;
            }
            else{
                high = mid - 1;
            }
        }
    
    return -1;
            
    }
}